### PR TITLE
Supercruise weapons, vision range, ion thrusters and proper shuttle propulsion

### PR DESCRIPTION
**Pull request by PowerfulBacon**
Its back and simpler than ever. I've taken out the NPCs, the mission changes and other crap. Here are just the weapons and back-end changes. Have fun.

## Administration

Recommended changes to space law:
 - Activating a GPS distress beacon in a non-emergency is a crime.

## About The Pull Request

~~Not a mega big update, but still an update none the less.~~
This is now a mega update, introducing a whole load of new features as well as shuttle weapon system from a long lost PR. #2551

### Shuttle Changes

All pilotable shuttles now require engines. The engines of a shuttle determine its acceleration properly and this will no longer be heavilly limited.

There are 2 new types of engine:
Ion Thruster: Consumes power instead of plasma.
Ion Burst Thruster: Consumes power instead of plasma, but will always consume more power than it can input so that it only runs temporarilly. Has a larger thrust ratio than regular thrusters.
Advanced Thruster: Premium engine for the nukie shuttle, cannot be destroyed and requires no fuel.

Shuttles now have a detection radius. This reduces the number of objects that need to be sent on the UI and limits how far a shuttle can see.

image

Shuttles can now undock from a location after they have initiated docking.

### Combat

Introducing the shuttle weapons console, a targeting console that allows for fierce combat with hostile ships.

Shuttle Weapons Console:
 - Will display all ships on the z-level
 - Select your weapon and click on the screen to fire.

image
image

### Bluespace Artillery

Nanotrasen's defence artillery used to strike enemy vessels can now be used to strike enemy vessels! Completing the Bluespace artillery will allow you to fire at any combat-capable shuttles, friend or foe.
image
image

### Distress Beacons

The GPS can now function as a distress beacon. When activated, the range of the orbital object the GPS is on will be extended to some big number so that everyone on the map can see where you are. Useful if you are stuck and you wish to inform someone to come and pick you up.
In future this could be expanded to have some interactions with NPC ships, however I couldn't figure out the behaviour of the NPCs once they are in mob form and the ship has docked.

image

### Misc Changes

Shuttle engines no longer accept any gas as fuel (Yes, you have been lied to, oxygen worked as fuel).
Tritium will now cause shuttle engines to have a 300% performance boost.

While docked at a station, the shuttle console will display the location of your shuttle rather than lavaland

### How things work

Shuttle Data: A single instance that exists per shuttle at all times the shuttle exists.
Shuttle Supercruise Object: An object representing the shuttle in flight. Exists while the shuttle is flying and on the map.
Shuttle AI Pilot: Attached to the shuttle data. Exists at all times the shuttle is controlled by an AI.

### Shuttle Data

Shuttle data is a way of tracking some data related to shuttles.
Shuttle data holds information about the stats of the shuttle, such as consumption rate, the engines attached to the shuttle as well as mass, health and some other things.
Shuttle data is created as a single instance per shuttle object.
To get the shuttle data attached to a shuttle call:
SSorbits.get_shuttle_data(**PORT_ID**)

## Why It's Good For The Game

This PR aims to heavily refactor a large portion of the mechanics related to supercruise/exploration, and change some of the worst parts of it.
The main thing being improved upon is the flight mechanics, at the moment there is nothing to do during flight and no skill involved from getting to A to B. After the changes, explorers will have to locate their target by using the machines they are provided, as well as avoid/fight hostile shuttles.
Secondly, this adds some more progression to explorers and some stuff that they rely on science for; Explorers do not start with any weapons on their ship so are very vulnerable if they get intercepted, and their only way to defend themselves is to board and take over the enemy ship. The progression factor is still being worked on, and technology disks will be improved.
VIP and assassination have been removed. This is because they led to boring interactions as the 2 parties were only interested in killing each other and nothing else. The ghost role potential will be transfered to NPC ships that fly around, which have no objectives and have no objectives to kill them. This means that any engagement between them and explorers is completely optional which opens up a lot more potential for interesting interactions/RP situations.
The goal of this PR is to make supercruise flight more engaging and fun, as well as add some new interesting content.

### An analysis of Exploration

Meaningful decisions are one of the most important parts of making engaging content. Currently exploration lacks meaningful decisions, its more of a process rather than a game.
A meaningful decision is one which:
 - The outcome of the decision has a meaningful impact.
 - A choice actually has to be made.

A choice where the outcome doesn't matter on the rest of the game is not a meaningful decision, a long with a choice where 1 option is clearly the best option.

Currently exploration consists of:
 - Selecting an objective
 - Completing the objective
 - Returning to the station with some money

This process has some dangers, and can be fun sometimes however lacks any grand importance. The main issue for this is me trying to make exploration have a limited impact on everything else, which completely goes against design for a game where every player is a small cog in a grand machine that falls apart when too many bad decisions are made - Every player should be able to impact the round, and a series of these impactful decisions is what leads to the chaotic nature of SS13, which in turn leads to fun rounds and good stories.

## Testing Photographs and Procedure

image

image

image

image

image

## Changelog
:cl: PowerfulBacon (Main Development), KubeRoot (Flight map UI improvements)
add: Added ion thruster
add: Added burst ion thruster
add: Added distress beacon on GPS
add: Shuttles can now undock from a location after they have initiated docking
add: Tritium can be used to turbo charge an engine, increasing performance by 300%
tweak: A bunch of shuttles have been editted to accomodate their new thrusters
tweak: Shuttles now have a detection radius
tweak: All drivable shuttles now require a proper engine source
fix: Shuttle engines no longer accept all gases
add: Shuttle weapon computer
add: Adds the Zeus MKI Railgun (Shuttle weapon)
add: Adds the Area Scattershot (Shuttle weapon)
add: Adds the Hades MkII Chaingun (Shuttle weapon)
add: Adds the Hades MkI Chaingun (Shuttle weapon)
add: Adds the Prometheus missile launcher (Shuttle weapon)
add: Adds the Minotaur missile launcher (Shuttle weapon)
add: Adds the Centaur II missile launcher (Shuttle weapon)
add: Adds the Centaur missile launcher (Shuttle weapon)
add: Adds the Zeus MKII Anti-Personnel Railgun (Shuttle weapon)
add: Adds the Burst laser (Shuttle weapon)
add: Adds the Burst laser MKII (Shuttle weapon)
add: Adds the Laser Cannon (Shuttle weapon)
add: BSA can now fire at shuttles.
add: New research tree for shuttle weaponry
add: Shuttles will now explode if they take too much damage
refactor: Refactors some shuttle code: AI pilot is now stored on the persistent shuttle_data class rather than on the shuttle object, meaning that shuttles can think even while not in flight.
balance: The listening outpost syndicate will now always spawn.
add: Shuttle autopilot will now automatically determine what thrust value to use in order to get smooth flight.
code: Refactors map loading callbacks to use priority queue rather than calling them in the order they were created.
code: Adds in the priority queue data structure.
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
I only made this PR for IDB and it didnt even run

**Comment by Fronsis** [comment]
Bless, in the provided screenshot i see that Explorers got a ''prisoner transport 16 objective'' does this mean Explorers got new objectives? or that is still being worked on

Also, can you target where you can shoot with the BSA/Weapons?

Really good stuff, i think it's missing the testing evidence of the space hazards? (i think it was ion/emp storm and something else from the old pr)

This could indeed have a lot of potential! I can imagine disks or buyable upgrades for the explorers to use to get weapons for the ship, really hope you can get down the road the NPCs ships with maybe some ghost roles to be random merchants and stuff like that, but i'm already liking that new type of objectives like prisoner shuttle, i hope people gave good suggestions and feedbacks on the player poll

**Comment by PowerfulBacon** [comment]
New objectives were removed (This was a copy from another, bigger PR)

**Comment by kit-katz** [comment]
finally... the starsector PR
EducatedNegativeAfricangoldencat-size_restricted

**Comment by EvilDragonfiend** [comment]
SS13 FTL WHEN

**Comment by PowerfulBacon** [comment]
This should be a relatively minor PR and not have too big design impacts.

**Comment by benbot16** [comment]
If the shields sprite is still a placeholder, you could probably steal NSVs FLARE sprite for it.

**Comment by Fronsis** [comment]
But what's the point of the shields if the hazard system during flight is still not introduced? Or do they help against incoming fire from enemy ships/bsa/etc? 

Do they also have an ''shield integrity'' bar as seen on the last picture? I think the testing evidence picture it's a little chaotic because it shows stuff that weren't ported from the big old pr

Sidenote: Man that npc objective looks so good, i hope it can get them introduced on the future, regardless thanks for this mini update!

**Comment by PowerfulBacon** [comment]
Yea, shields weren't continued since they were useless in testing

**Comment by WrillWasTaken** [comment]
it takes AGES to land ANYTHING, it was a solid 5 minute waiting to land on a custom location on the station z level. it also took 5 minutes to dock to a procedural ruin.

**Comment by PowerfulBacon** [comment]
it took forever to land due to a runtime, admins were manually sending shuttles

**Comment by Crossedfall** [comment]
> SS13 FTL WHEN

Four years ago https://github.com/FTL13/FTL13

**Comment by PowerfulBacon** [comment]
Admin spawned ships seem to start with 0 health and are instantly destroyed

**Comment by PowerfulBacon** [comment]
Shuttles fail to dock

**Comment by PowerfulBacon** [comment]
If a target is no longer on orbital map, triangulation computer should say something rather than do nothing

**Comment by PowerfulBacon** [comment]
Triangulation computer just needs a rework

**Comment by PowerfulBacon** [comment]
> Soo, you can still shoot at ships even if they are out of your targeting range if you targeted them before.

**Comment by PowerfulBacon** [comment]
Interdiction screen crashes the UI

**Comment by Cenrus** [comment]
Helm console bluescreened:

```NotFoundError
   at http://127.0.0.1:51520/tmp2632/tgui.bundle.js:1:174380
User Agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.2; WOW64; Trident/7.0; .NET4.0C; .NET4.0E; .NET CLR 2.0.50727; .NET CLR 3.0.30729; .NET CLR 3.5.30729)
State: {"ckey":"cenrus","interface":"OrbitalMap","window":{"key":"[0x2030993]-main","size":null,"fancy":1,"locked":0}}

**Comment by PowerfulBacon** [comment]
Explo ship lost their locator console (refactor this)

**Comment by PowerfulBacon** [comment]
- Weapons still fire backwards
- You can't manually link weapons console
- Shuttles with dynamic IDs instantly break
- The local map on the shuttle consoles doesn't work
- The locator console is too confusing for people
- The research and circuitboard is missing for the locator console

**Comment by PowerfulBacon** [comment]
The locator console has been updated to be slightly easier to use and hopefully less confusing.
image
It will display possible positions of the target, the more scans you do, the more accurate they become.

**Comment by PowerfulBacon** [comment]
Issue: Local map on shuttle consoles can see stealthed ships (infiltrator)

File: _maps/_basemap.dm [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//#define LOWMEMORYMODE //uncomment this to load centcom and runtime station and thats it.
+#define LOWMEMORYMODE //uncomment this to load centcom and runtime station and thats it.
```
Comment (PowerfulBacon) [inline _maps/_basemap.dm]:
```suggestion
//#define LOWMEMORYMODE //uncomment this to load centcom and runtime station and thats it.
```

Comment (PowerfulBacon) [reply _maps/_basemap.dm]: Really should have spotted that one

**Comment by PowerfulBacon** [comment]
Note: There should be a way to add waypoints on the locator console that the pilot can see

**Comment by CydiaLamiales** [comment]
FTL in SS13, by body is ready.

**Comment by PowerfulBacon** [comment]
Weapons dont get removed from consoles when deleted

**Comment by wikimody** [comment]
First thank you for making this, my bigest dream in SS13 is making functional batelcruser and you enebled me to actuali do it.
Now a qustion: How do wepons interact with 3 shield systems curently present? Those are anti breach shields (not sure if those work in traisit), emiter powerd shields (the ones used in engieniering for teslas and singularities, bouncy) and shield wals (the ones from xenobio).

**Comment by PowerfulBacon** [comment]
Anti-breach shielding projectors will block projectiles and be destroyed.
The other 2 I believe will have the projectiles pass through them but I'm not sure.

**Review by PestoVerde322 — DISMISSED** [review]
Maps looks fine, icons too, Approved

**Comment by wikimody** [comment]
When will this start testmerging? I'm exited for this to be implemented.

**Comment by PowerfulBacon** [comment]
This has already gone through testmerging

**Comment by Bastian0930** [comment]
image
I'm unsure if this is related to the test merge, but I just had this happen in my first round with the merge. I just docked next to a window after waiting 20 seconds to make sure it was done loading.

**Comment by PowerfulBacon** [comment]
Ah whoops, that was an issue with slow map placing

**Comment by LodedDiper69** [comment]
Help, I found a bug on a TM where space cops cant use their shuttle! That needs to be fixed.

**Comment by PowerfulBacon** [comment]
Players can land at ruins before they are fully generated still, which is strange because that should not happen

**Comment by PowerfulBacon** [comment]
Back to the annual stale realm

**Comment by CydiaLamiales** [comment]
NOOOOOO

**Pull request closed without merging**